scrape/manager_test.go

```diff
@@ -703,3 +710,79 @@ scrape_configs:
 	reload(scrapeManager, cfg2)
 	require.ElementsMatch(t, []string{"job1", "job3"}, scrapeManager.ScrapePools())
 }
+
+func TestManagerScrapeCreatedTimestamp(t *testing.T) {
+	app := &collectResultAppender{}
+	scrapeManager := NewManager(&Options{
+		EnableProtobufNegotiation: true,
+	}, log.NewLogfmtLogger(os.Stderr), &collectResultAppendable{app})
+
+	require.NoError(t, scrapeManager.ApplyConfig(&config.Config{
+		GlobalConfig: config.GlobalConfig{
+			// Extremely high scrape interval, to ensure the only chance to see the
+			// sample is on start and stopAfterScrapeAttempt.
+			ScrapeInterval: model.Duration(5 * time.Second),
+			ScrapeTimeout:  model.Duration(5 * time.Second),
+		},
+		ScrapeConfigs: []*config.ScrapeConfig{{JobName: "test"}},
+	}))
+
+	metric := &dto.MetricFamily{
+		Name: proto.String("expected_metric"),
+		Metric: []*dto.Metric{
+			{
+				Counter: &dto.Counter{
+					Value:            proto.Float64(1.0),
+					CreatedTimestamp: &timestamppb.Timestamp{Seconds: 1, Nanos: 1},
+				},
+			},
+		},
+	}
+
+	b, err := proto.Marshal(metric)
+	require.NoError(t, err)
+
+	// Start fake HTTP target to scrape returning a single metric.
+	server := httptest.NewServer(
+		http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			w.Header().Set("Content-Type", `application/vnd.google.protobuf; proto=io.prometheus.client.MetricFamily; encoding=delimited`)
+			w.Write(b)
+		}),
+	)
+	defer server.Close()
+
+	serverURL, err := url.Parse(server.URL)
+	require.NoError(t, err)
+
+	// Add fake target directly into tsets + reload. Normally users would use
+	// Manager.Run and wait for minimum 5s refresh interval.
+	scrapeManager.updateTsets(map[string][]*targetgroup.Group{
+		"test": {
+			{
+				Targets: []model.LabelSet{{
+					model.SchemeLabel:  model.LabelValue(serverURL.Scheme),
+					model.AddressLabel: model.LabelValue(serverURL.Host),
+				}},
+			},
+		},
+	})
+	scrapeManager.reload()
+	// Wait for one scrape
+	time.Sleep(5 * time.Second)
```

Comment:
Let's remove sleep from tests - for all costs (:

Use Retry method instead

Reply:
Note: Bartek's retry: https://github.com/prometheus/prometheus/pull/12687/files#diff-63477d3794c3d48b17198c56aa49e399501bb438c91057d47db8e263d7462975

Although, I'm surprised there isn't a utility for this already in one of our third-party libraries. k8s libraries provide something similar.